storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateEntry.java

```diff
@@ -0,0 +1,143 @@
+        this.coordinatorEpoch = coordinatorEpoch;
+        this.lastTimestamp = lastTimestamp;
+        this.currentTxnFirstOffset = currentTxnFirstOffset;
+    }
+
+    public int firstSeq() {
+        return isEmpty() ? RecordBatch.NO_SEQUENCE : batchMetadata.get(0).firstSeq();
+    }
+
+
+    public long firstDataOffset() {
+        return isEmpty() ? -1L : batchMetadata.get(0).firstOffset();
+    }
+
+    public int lastSeq() {
+        return isEmpty() ? RecordBatch.NO_SEQUENCE : batchMetadata.get(batchMetadata.size() - 1).lastSeq;
+    }
+
+    public long lastDataOffset() {
+        return isEmpty() ? -1L : batchMetadata.get(batchMetadata.size() - 1).lastOffset;
+    }
+
+    public int lastOffsetDelta() {
+        return isEmpty() ? 0 : batchMetadata.get(batchMetadata.size() - 1).offsetDelta;
+    }
+
+    public boolean isEmpty() {
+        return batchMetadata.isEmpty();
+    }
+
+    public void addBatch(short producerEpoch, int lastSeq, long lastOffset, int offsetDelta, long timestamp) {
+        maybeUpdateProducerEpoch(producerEpoch);
+        addBatchMetadata(new BatchMetadata(lastSeq, lastOffset, offsetDelta, timestamp));
+        this.lastTimestamp = timestamp;
+    }
+
+    public boolean maybeUpdateProducerEpoch(short producerEpoch) {
+        if (this.producerEpoch != producerEpoch) {
+            batchMetadata.clear();
+            this.producerEpoch = producerEpoch;
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private void addBatchMetadata(BatchMetadata batch) {
+        if (batchMetadata.size() == ProducerStateEntry.NUM_BATCHES_TO_RETAIN) batchMetadata.remove(0);
+        batchMetadata.add(batch);
+    }
+
+    public void update(ProducerStateEntry nextEntry) {
+        maybeUpdateProducerEpoch(nextEntry.producerEpoch);
+        while (!nextEntry.batchMetadata.isEmpty()) addBatchMetadata(nextEntry.batchMetadata.remove(0));
+        this.coordinatorEpoch = nextEntry.coordinatorEpoch;
+        this.currentTxnFirstOffset = nextEntry.currentTxnFirstOffset;
+        this.lastTimestamp = nextEntry.lastTimestamp;
+    }
+
+    public Optional<BatchMetadata> findDuplicateBatch(RecordBatch batch) {
+        if (batch.producerEpoch() != producerEpoch) return Optional.empty();
+        else return batchWithSequenceRange(batch.baseSequence(), batch.lastSequence());
+    }
+
+    // Return the batch metadata of the cached batch having the exact sequence range, if any.
+    Optional<BatchMetadata> batchWithSequenceRange(int firstSeq, int lastSeq) {
+        Stream<BatchMetadata> duplicate = batchMetadata.stream().filter(metadata -> firstSeq == metadata.firstSeq() && lastSeq == metadata.lastSeq);
+        return duplicate.findFirst();
+    }
+
+    public List<BatchMetadata> batchMetadata() {
+        return Collections.unmodifiableList(batchMetadata);
+    }
+
+    public short producerEpoch() {
+        return producerEpoch;
+    }
+
+    @Override
+    public String toString() {
```

Comment:
This implementation differs from the original one. Is there a reason for that?

Reply:
It was a miss from my side. I generated using intellij and forgot to check with the earlier Scala toString method.